chain/chain/src/flat_storage_creator.rs

```diff
@@ -1,47 +1,109 @@
     #[allow(unused)]
-    pub traversed_parts_receiver: Receiver<u64>,
+    fetched_parts_receiver: Receiver<u64>,
 }
 
 impl FlatStorageShardCreator {
-    pub fn new(status: FlatStorageStateStatus, shard_id: ShardId) -> Self {
-        let (traversed_parts_sender, traversed_parts_receiver) = unbounded();
+    pub fn new(
+        shard_id: ShardId,
+        start_height: BlockHeight,
+        runtime_adapter: Arc<dyn RuntimeAdapter>,
+    ) -> Self {
+        let (fetched_parts_sender, fetched_parts_receiver) = unbounded();
         Self {
-            status,
             shard_id,
-            traversed_state_parts: None,
-            traversed_parts_sender,
-            traversed_parts_receiver,
+            start_height,
+            runtime_adapter,
+            fetched_state_parts: None,
+            fetched_parts_sender,
+            fetched_parts_receiver,
+        }
+    }
+
+    #[cfg(feature = "protocol_feature_flat_state")]
+    pub(crate) fn update_status(&mut self, chain_store: &ChainStore) -> Result<(), Error> {
+        let current_status =
+            store_helper::get_flat_storage_state_status(chain_store.store(), self.shard_id);
+        match current_status {
+            FlatStorageStateStatus::SavingDeltas => {
+                let final_head = chain_store.final_head()?;
+                let shard_id = self.shard_id;
+
+                if final_head.height > self.start_height {
+                    // If it holds, deltas for all blocks after final head are saved to disk, because they have bigger
+                    // heights than one on which we launched a node. Check that it is true:
+                    for height in final_head.height + 1..=chain_store.head()?.height {
+                        for (_, hashes) in
+                            chain_store.get_all_block_hashes_by_height(height)?.iter()
+                        {
+                            for hash in hashes {
+                                debug!(target: "chain", %shard_id, %height, %hash, "Checking delta existence");
+                                assert_matches!(
+                                    store_helper::get_delta(
+                                        chain_store.store(),
+                                        shard_id,
+                                        hash.clone(),
+                                    ),
+                                    Ok(Some(_))
+                                );
+                            }
+                        }
+                    }
+
+                    // We continue saving deltas, and also start fetching state.
+                    let block_hash = final_head.last_block_hash;
+                    let mut store_update = chain_store.store().store_update();
+                    store_helper::set_flat_head(&mut store_update, shard_id, &block_hash);
+                    store_helper::set_fetching_step(&mut store_update, shard_id, 0u64);
```

Comment:
What is fetching step?

Reply:
"Fetching state step" is probably a better name here, it is described in comment for `FlatStorageStateStatus::FetchingState`.
It means the state part which is currently being written to flat storage.